source/MRViewer/MRSceneCache.h

```diff
@@ -16,23 +15,36 @@ class MRVIEWER_CLASS SceneCache
     // call it in the beginning each frame
     MRVIEWER_API static void invalidateAll();
 
-    // same as getAllObjectsInTree<ObjectType>( &SceneRoot::get(), SelectivityType ) but use cached data
+    template <typename ObjectType>
+    using TemplateDataType = std::vector<std::shared_ptr<ObjectType>>;
+    // analog getAllObjectsInTree<ObjectType>( &SceneRoot::get(), SelectivityType ) but use cached data
+    // reference is valid until invalidateAll() is called
     template <typename ObjectType, ObjectSelectivityType SelectivityType>
-    static const std::vector<std::shared_ptr<ObjectType>>& getAllObjects();
+    static const TemplateDataType<ObjectType>& getAllObjects();
 
+    template <typename ObjectType>
+    using CachedTemplateDataType = std::shared_ptr<TemplateDataType<ObjectType>>;
+    // analog getAllObjectsInTree<ObjectType>( &SceneRoot::get(), SelectivityType ) but use cached data
+    // shared ownership of cached data for current frame( not updated for new frame )
+    template <typename ObjectType, ObjectSelectivityType SelectivityType>
+    static CachedTemplateDataType<ObjectType> getAllObjectsCached();
+
+    using DepthDataType = std::vector<int>;
```

Comment:
please add a comment what is inside this vector

Reply:
reworked